nixos/kubernetes: use nix coredns package by default

johanot opened the pull request:
The coredns package uses the kubernetes dns tests as passthru, but currently the coredns image is fetched upstream from Docker Hub by default. So actually, the coredns nix package is not being tested.

I believe it is sensible to use the Nix built CoreDNS package by default. If users want to pull the upstream image instead, they can do so by setting `services.kubernetes.addons.dns.corednsImage`.

supersedes #418071

cc
@MisterRager
@DeltaEvo
@djds
@rtreffer
@rushmorem

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [X] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [X] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

johanot commented:
@djds I didn't really look at the upstream image tbh. I only checked that the Kubernetes dns-tests still pass with the NixOS image. But the upstream Dockerfile is here: https://github.com/coredns/coredns/blob/master/Dockerfile

I can see that it has ca-certificates and libcap. I don't know for what usecase Coredns would need that? But CoreDNS is kind of a swiss army knife. We can add pkgs.libcap and pkgs.cacert? What do you think? :) IMHO... As the _default_ for nixos/kubernetes, this is fine. Users can always override the with a custom image if they need more features.

johanot commented:
@saschagrunert you might have an opinion here? :)

saschagrunert reviewed (approved):
Thanks, yeah LGTM overall.

johanot commented:
Alternatively, if this is not merged, the passthru.tests of coredns should be removed. They signal false security for the coredns package atm.

johanot committed:
nixos/kubernetes: use nixos coredns package per default

- services.kubernetes.addons.dns.coredns is now
services.kubernetes.addons.dns.corednsImage
- the default value of the option is now a simple
dockerTools-buildImage
- added rel note for nixos 26.05

johanot commented:
rel note moved to nixos 26.05. Could still be nice to get this merged. :pray:

srhb commented:
Sorry I'm late to this. Going by the various maintainer approvals in the thread, let me run the tests on a fresh base and click the button if they pass.
